### PR TITLE
Copy code from Bert to Roberta and add safeguard script

### PR DESCRIPTION
**Pull request by sgugger**
This PR proposes some tooling to automatically check copied code from one of the files in transformers to another is consistent with the original. To showcase an example, it uses `modeling_roberta.py` where I copied all code from `modeling_bert.py` instead of importing the `BertModel` (as a result only activations are imported from `modeling_bert` and an additional refactor to put them in `activations` would make the file fully independent from `modeling_bert` but that is not the scope of this PR).

A new script called `check_copies` is introduced in the utils folder. It will check all code files for comments looking like:
```python
# Copied from transformers.modeling_bert.BertEmbeddings
```
When it encounters such a comment, it will recover the code of the object (here `BertEmbeddings`) inside its module (here `modeling_bert`) and compare it to the code in the class/function where this comment is. If there is a difference, the script will replace and overwrite (in overwrite mode) or raise an error (in checking mode). This way we can be aware of any difference when a fix is made in bert and should be ported to roberta, and we can automatically copy that difference.

The script works for classes, functions and methods. It can be called inside a function/method (if we want to add some specific initial code, see some examples below). I've tried to make the comments readable while ensuring some non-related comments could be caught (hence the `transformers.` that is not useful information),

It also has a functionality where it can replace one pattern by another, for instance:
```python
# Copied from transformers.modeling_bert.BertEmbeddings with Bert->Roberta
```
will look for a diff between the original code and the observed code after replacing all instance of Bert in the original code by Roberta (this is because we have `BertLayerNorm`, `BertPooler`, `BertEncoder`... that becomes `RobertaLayerNorm`, `RobertaPooler`, `RobertaEncoder`...). Again I tried to aim for readable comments that could still be used for the script.

I added the script in overwriting mode in `make style` and in checking mode in `make quality`. Users have to be made aware that `make style` can make destructive changes when it's run (since it will erase any modification in copied parts). I can create a new alias if we think this is too dangerous for `make style`.

### CODE REVIEW

File: src/transformers/modeling_roberta.py [hunk 1]
```diff
@@ -57,15 +65,31 @@
...
+
+        # self.LayerNorm is not snake-cased to stick with TensorFlow model variable name and be able to load
+        # any TensorFlow checkpoint file
+        self.LayerNorm = RobertaLayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+
+        # position_ids (1, len position emb) is contiguous in memory and exported when serialized
+        self.register_buffer("position_ids", torch.arange(config.max_position_embeddings).expand((1, -1)))
+
+        # End copy
```
Comment (sgugger) [inline src/transformers/modeling_roberta.py:91]: This special comment tells the script to stop looking at the code at this point, allowing us to have custom stuff after. This is the equivalent of doing
```
super().__init__(config)
subclass_specific_stuff
```

File: src/transformers/modeling_roberta.py [hunk 2]
```diff
@@ -78,9 +102,29 @@ def forward(self, input_ids=None, token_type_ids=None, position_ids=None, inputs
             else:
                 position_ids = self.create_position_ids_from_inputs_embeds(inputs_embeds)
 
-        return super().forward(
-            input_ids, token_type_ids=token_type_ids, position_ids=position_ids, inputs_embeds=inputs_embeds
-        )
+        # Copied from transformers.modeling_bert.BertEmbeddings.forward
```
Comment (sgugger) [inline src/transformers/modeling_roberta.py:105]: The other way round, we can add any custom code at the beginning, *then* compare to the original code, by just putting the comment later in the function body and not at the beginning.

The only thing that is impossible to do is to change the middle of the copied code.

File: src/transformers/modeling_roberta.py [hunk 3]
```diff
@@ -98,6 +142,343 @@ def create_position_ids_from_inputs_embeds(self, inputs_embeds):
...
+
+        context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
+        new_context_layer_shape = context_layer.size()[:-2] + (self.all_head_size,)
+        context_layer = context_layer.view(*new_context_layer_shape)
+
+        outputs = (context_layer, attention_probs) if output_attentions else (context_layer,)
+        return outputs
+
+
+# Copied from transformers.modeling_bert.BertSelfOutput with Bert->Roberta
```
Comment (sgugger) [inline src/transformers/modeling_roberta.py:224]: Example of patter with a replacement (`BertLayerNorm` in the original file, `RobertaLayerNorm` here).

File: src/transformers/modeling_roberta.py [hunk 4]
```diff
@@ -29,15 +29,23 @@
     add_start_docstrings_to_callable,
     replace_return_docstrings,
 )
-from .modeling_bert import BertEmbeddings, BertLayerNorm, BertModel, BertPreTrainedModel, gelu
+from .modeling_bert import ACT2FN, gelu
```
Comment (LysandreJik) [inline src/transformers/modeling_roberta.py:32]: While we're at it, this can be imported from `activations`

File: src/transformers/modeling_roberta.py [hunk 5]
```diff
@@ -98,6 +142,343 @@ def create_position_ids_from_inputs_embeds(self, inputs_embeds):
...
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, all_hidden_states, all_attentions] if v is not None)
+        return BaseModelOutput(
+            last_hidden_state=hidden_states, hidden_states=all_hidden_states, attentions=all_attentions
+        )
+
+
+# Copied from transformers.modeling_bert.BertPooler
```
Comment (LysandreJik) [inline src/transformers/modeling_roberta.py:443]: Should the replacement suggestion be done for the class name as well, or only if its internals has the name in it?
```suggestion
# Copied from transformers.modeling_bert.BertPooler with Bert->Roberta
```

File: src/transformers/modeling_roberta.py [hunk 6]
```diff
@@ -180,14 +576,121 @@ def get_input_embeddings(self):
...
+        heads_to_prune: dict of {layer_num: list of heads to prune in this layer}
+        See base class PreTrainedModel
+        """
+        for layer, heads in heads_to_prune.items():
+            self.encoder.layer[layer].attention.prune_heads(heads)
+
+    @add_start_docstrings_to_callable(ROBERTA_INPUTS_DOCSTRING.format("(batch_size, sequence_length)"))
+    @add_code_sample_docstrings(
+        tokenizer_class=_TOKENIZER_FOR_DOC,
+        checkpoint="robert-base-uncased",
```
Comment (LysandreJik) [inline src/transformers/modeling_roberta.py:590]:
```suggestion
        checkpoint="roberta-base",
```

**Review by LysandreJik — COMMENTED** [review]
This is a very nice addition, it brings much more safety when copy-pasting! I'm 100% in favor of this change; this would make maintaining the model files simpler while removing literally nothing from the user experience.

Two things I think we should clear up before merging:
- Could it be tested?
- It is run after `black`. In this case you're showcasing the change from BERT to RoBERTa, but if we were to change it to a very long imaginary name like `BertForGenerationAndComputerVisionBoundingBoxesForVisualQuestionAnswering` (can't wait to have it in the lib!), then `black` would probably have to reformat some of the resulting file. When comparing the two files, your script wouldn't be able to understand that, right?

File: src/transformers/modeling_roberta.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/modeling_roberta.py:443]: It doesn't check the class name, only the code inside (unless it's a method, the name of the class function is expected to change from the original).

**Comment by sgugger** [comment]
- For the tests, I'm not sure what you had in mind. I could create a dumb file that I then compare to the lib.
- It does not matter if this is run before or after `black`: if there is a very long model name, the script will detect differences as soon as black as indeed been run (so it may pass if run before black once but fail at the next CI). Running it after makes sure we don't catch trailing spaces or bad empty lines (empty lines with spaces) in the diff. But this all won't work for replacement of short model names to very long model names (so sadly, we can't have `BertForGenerationAndComputerVisionBoundingBoxesForVisualQuestionAnswering` in the lib).

**Comment by LysandreJik** [comment]
- Regarding the tests, I think it would be a requirement given that the script modifies the code when ran with `overwrite`. Testing to ensure it works correctly seems necessary.
- I think it could work with a bit of plumbing: given that the script registers the name change `Bert->Roberta`, the script could change the copied file from `Roberta` to `Bert`, run `black` to correctly format, and check equivalence then. I haven't though much about edge cases but it seems robust enough to me. What do you think?

**Comment by patrickvonplaten** [comment]
This is great!

**Review by LysandreJik — APPROVED** [review]
Great, thanks for adding the tests!
